splunk/SplunkUtills/SplunkUtills.py

```diff
@@ -0,0 +1,150 @@
+
+    Kwargs:
+        columns (list): columns to select in the query. If None, select all columns.
+        max_columns (int): maximum number of rows in the query. If None, the query will have no rows limitation.
+
+        Additional kwargs for filtering by column value.
+        Note that query filter order is based on this kwargs order, except for the indexes of "index", "earliest" & "latest".
+
+    Returns:
+        String: The SPL query.
+    '''
+
+    # transform query_kwargs to params and hyperparams
+    hyperparams = ['max_columns', 'columns']
+
+    indexes_order = ['index', 'earliest', 'latest']
+
+    params = [param for param in indexes_order if param in query_kwargs.keys()] + \
+             [param for param in query_kwargs.keys() if param not in indexes_order and param not in hyperparams]
+
+    # filter rows ("where")
+    query = "search "
+
+    where_list = []
+
+    for param in params:
+        if query_kwargs[param]:
+            if isinstance(query_kwargs[param], (list, tuple,dict, set)):
+                print("can't filter by list ({})\n".format(param))
+            else:
+                where_list.append('{}={} '.format(param, query_kwargs[param]))
+
+    if where_list:
+        query = query + "".join(where_list) + '|'
+
+    else:
+        query = query + ' *|'
+
+    # filter columns ("select")
+    if 'columns' in query_kwargs and isinstance(query_kwargs['columns'], list):
+        query = query + " fields " + ", ".join(query_kwargs['columns']) + " |"
+    else:
+        query = query + ' fields * |'
+
+    # cut rows ("top")
+    if 'max_columns' in query_kwargs:
+        query = query + ' head {}'.format(query_kwargs['max_columns'])
+
+    return query
+
+
+
+def read_splunk(query, username, password, host, port):
+    '''This function queries Splunk database.
+    Note that function dependencies includes installing splunklib.
+
+    Args:
+        query (str): SPL query.
+        username (str): Splunk username.
+        password (str): Splunk password.
+        host (str): Splunk host.
+        port (int): Splunk port.
+
+    Returns:
+        DataFrame: query results.
+    '''
+
+    print('start..\n')
+
+    print('your query:\n {}\n'.format("|\n".join(query.split("|"))))
+
+    # imports
+    import sys
+    import re
+    import pandas as pd
+    from io import BytesIO
+    from time import sleep
+
+    if 'splunk-sdk-python-1.6.6' not in sys.path:
+        try:
+            sys.path.append('splunk-sdk-python-1.6.6')
+        except:
+            None
+
+    import splunklib.client as client
+
+    print('imports completed\n')
+
+    # connect to splunk
+    service = client.connect(host=host, port=port, username=username, password=password)
+    print('connection succeed\n')
+
+    # query splunk
+    kwargs_normalsearch = {"exec_mode": "normal", "count": 0}
+    job = service.jobs.create(query, **kwargs_normalsearch)
+

```

Comment:
I think we should add a time limit (with some reasonable default, a few minutes), and cancel the job when it is reached.

Reply:
Done